src/cis/fam_cis_direct.cpp

```diff
@@ -248,25 +257,61 @@ Fam_CIS_Direct::create_region(string name, size_t nbytes, mode_t permission,
     CIS_DIRECT_PROFILE_START_OPS()
     Fam_Region_Metadata region;
     ostringstream message;
+    int user_policy = 0;
+    uint64_t regionId = -1;
+    std::list<int> memory_server_list;
+    uint64_t *memServerIds;
+    int used_memsrv_cnt = 0;
 
-    uint64_t memoryServerId = generate_memory_server_id(name.c_str());
+    // List of memory servers where a given region need to be spanned.
+    std::list<Fam_Memory_Service *> memoryServiceList;
 
-    Fam_Memory_Service *memoryService = get_memory_service(memoryServerId);
-    Fam_Metadata_Service *metadataService =
-        get_metadata_service(memoryServerId);
-    // Check if the name size is bigger than MAX_KEY_LEN supported
-    if (name.size() > metadataService->metadata_maxkeylen()) {
-        message << "Name too long";
-        THROW_ERRNO_MSG(CIS_Exception, REGION_NAME_TOO_LONG,
-                        message.str().c_str());
-    }
+    // TODO: For now we are using 0 as id for metadata server.
+    // This piece of code may be revisted, if needed.
+    Fam_Metadata_Service *metadataService = get_metadata_service(0);
 
-    if (metadataService->metadata_find_region(name, region)) {
-        message << "Region already exist";
-        THROW_ERRNO_MSG(CIS_Exception, REGION_EXIST, message.str().c_str());
+    // Validate that region can be created and get regionId and memory servers
+    // in which region needs to be spanned.
+    try {
+        metadataService->metadata_validate_and_create_region(
+            name, nbytes, &regionId, &memory_server_list, user_policy);
+    } catch (...) {
+        throw;
+    }
+    // Code for spanning region across multiple memory servers.
+    // Asyncronously create regions in multiple memory servers for given region
+    // id.
+    memServerIds =
+        (uint64_t *)malloc(sizeof(uint64_t) * memory_server_list.size());
+    for (auto it = memory_server_list.begin(); it != memory_server_list.end();
+         ++it) {
+        memoryServiceList.push_back(get_memory_service(*it));
+        memServerIds[used_memsrv_cnt++] = *it;
+    }
+
+    std::list<std::shared_future<void>> resultList;
+    try {
+    for (auto memsrv : memoryServiceList) {
+        Fam_Memory_Service *memoryService = memsrv;
+        std::future<void> result(std::async(
+            std::launch::async, &openfam::Fam_Memory_Service::create_region,
+            memoryService, regionId, nbytes/used_memsrv_cnt));
+        resultList.push_back(result.share());
+    }
+    } catch (...) {
+        metadataService->metadata_reset_bitmap(regionId);
+        throw;
     }
 
-    uint64_t regionId = memoryService->create_region(nbytes);
+    // Wait for region creation to complete.
+    try {
+        for (auto result : resultList) {
+            result.get();
+        }
+    } catch (...) {
+        metadataService->metadata_reset_bitmap(regionId);
```

Comment:
In addition to reseting the bitmap, we need to revert back the created region in some memory servers. i.e call destroy region, where ever create_region was successful.